drivers: wifi: esp_at: fix some crash issues

ccli8 opened the pull request:
This fixes some crash issues on `esp_at` modem when running `net/sockets/big_http_download` sample with tls enabled, including:
1. Premature socket release when `esp_rx` thread can still hold reference to this socket
2. Deadlock in socket close path

ccli8 committed:
drivers: wifi: esp_at: fix premature socket release

This fixes premature socket release. Without this fix, system will crash
because esp_rx thread can still hold reference to the socket in the
following cases:
1. esp_rx thread has got the socket reference but hasn't unref'ed
   via esp_socket_unref.
2. esp_rx thread can still get the socket reference for refcount not
   being zero in esp_socket_ref.

Signed-off-by: Chun-Chieh Li <ccli8@nuvoton.com>

ccli8 committed:
drivers: wifi: esp_at: fix deadlock in socket close path

This fixes deadlock in socket close path. In the scenario:
1. on_cmd_ipd/esp_socket_rx invokes esp_socket_ref_from_link_id
   and increments refcount.
2. zvfs_close/esp_put locks cond.lock.
3. zvfs_close/esp_put waits on sem_free.
4. on_cmd_ipd/esp_socket_rx waits on cond.lock before esp_socket_unref.
5. sem_free waits on esp_socket_unref for refcount reaching zero.

As we detect socket is closing, we can ignore last rx data and escape
from the trap.

Signed-off-by: Chun-Chieh Li <ccli8@nuvoton.com>

jukkar commented on drivers/wifi/esp_at/esp_socket.c:
```diff
@@ -76,7 +76,10 @@ void esp_socket_unref(struct esp_socket *sock)
 	} while (!atomic_cas(&sock->refcount, ref, ref - 1));
 
-	k_sem_give(&sock->sem_free);
+	/* notifies free only on 1-to-0 transition */
+	if (ref == 1) {
+		k_sem_give(&sock->sem_free);
+	}
```
I suggest using the same pattern like in other parts of the system (for example in `net_pkt.c`) so it is easier to understand what is going on
```
if (ret > 1) {
	return;
}

k_sem_give(&sock->sem_free);
```

ccli8 replied:
follow it

jukkar commented on drivers/wifi/esp_at/esp_socket.c:
```diff
@@ -192,8 +192,30 @@ void esp_socket_rx(struct esp_socket *sock, struct net_buf *buf,
+		 * If the socket is closing, we can ignore the packet and won't
+		 * trap in deadlock.
+		 */
+		while ((ref = atomic_get(&sock->refcount)) > 1 && ret == -EAGAIN) {
```
The ref variable does not seem to be used for anything, do we really need it?
Why not
```
while (atomic_get(&sock->refcount) > 1 && ret == -EAGAIN) {
```

ccli8 replied:
removed `ref` variable

ccli8 commented on drivers/wifi/esp_at/esp_socket.c:
```diff
@@ -192,8 +192,30 @@ void esp_socket_rx(struct esp_socket *sock, struct net_buf *buf,
+		while ((ref = atomic_get(&sock->refcount)) > 1 && ret == -EAGAIN) {
+			ret = k_mutex_lock(sock->context->cond.lock, K_SECONDS(1));
+		}
+		if (ret) {
```
thanks for informing this rule, follow it

ccli8 committed:
drivers: wifi: esp_at: fix rx_sock not ref-counted

This fixes rx_sock is not reference-counted, or crash error on reference
to released socket.

Signed-off-by: Chun-Chieh Li <ccli8@nuvoton.com>

73jn commented on drivers/wifi/esp_at/esp_socket.c:
```diff
@@ -189,8 +194,29 @@ void esp_socket_rx(struct esp_socket *sock, struct net_buf *buf,
+		while (atomic_get(&sock->refcount) > 1 && ret == -EAGAIN) {
+			ret = k_mutex_lock(sock->context->cond.lock, K_SECONDS(1));
+		}
+		if (ret != 0) {
+			/* Discard */
+			net_pkt_unref(pkt);
+			return;
+		}
```
Thanks for your feedback. I've tested your patch and everything's resolved. I owe you a beer!

marekmatej reviewed (approved):
lgtm
